### PR TITLE
✨ Feat/open margin account

### PR DESCRIPTION
**Pull request by Thomas-Smets**
- Create interface between Arcadia Vaults and Protocols in need of margin accounts
- Clean up Vault code

### CODE REVIEW

File: src/Vault.sol [hunk 1]
```diff
@@ -32,48 +33,32 @@ import "./interfaces/ILiquidityPool.sol";
...
-    // ACCESS CONTROL
-    address public owner;
-    mapping(address => bool) public allowed;
-
-
     // Each vault has a certain 'life', equal to the amount of times the vault is liquidated.
     // Used by the liquidator contract for proceed claims
     uint256 public life;
-
     bool public initialized;
```
Comment (j-vp) [inline src/Vault.sol:52]: I believe the initialized var can be removed, init is now checked with vault version.

File: src/Vault.sol [hunk 2]
```diff
@@ -658,92 +648,113 @@ contract Vault {
...
+    @dev First checks if there is no locked value. If there is no value locked then the baseCurrency gets changed to the param
   */
     function _setBaseCurrency(
-        uint256 newBaseCurrency
+        address _baseCurrency
     ) private {
-        require(getUsedMargin() == 0, "VL: Can't change baseCurrency when openDebt > 0");
-        require(newBaseCurrency + 1 <= IMainRegistry(_registryAddress).baseCurrencyCounter(), "VL: baseCurrency not found");
-        vault._baseCurrency = uint8(newBaseCurrency); //Change this to where ever it is going to be actually set
+        require(getUsedMargin() == 0, "VL: Can't change baseCurrency when Used Margin > 0");
```
Comment (j-vp) [inline src/Vault.sol:671]: Add function after VL

File: src/Vault.sol [hunk 3]
```diff
@@ -658,92 +648,113 @@ contract Vault {
...
+        require(IMainRegistry(registryAddress).isBaseCurrency(_baseCurrency), "VL: baseCurrency not found");
```
Comment (j-vp) [inline src/Vault.sol:672]: Same

File: src/Vault.sol [hunk 4]
```diff
@@ -658,92 +648,113 @@ contract Vault {
...
+    function openTrustedMarginAccount(address protocol) onlyOwner public {
+        require(!isTrustedProtocolSet, "V_OMA: ALREADY SET");
+        //ToDo: Check in Factory/Mainregistry if protocol is indeed trusted?
 
-        for (uint256 i = 0; i < len; ) {
-            // optimizooooor
-            assembly {
-                sum := add(sum, mload(add(add(_data, 0x20), mul(i, 0x20))))
-            }
+        (bool success, address baseCurrency, address liquidator_) = ITrustedProtocol(protocol).openMarginAccount();
```
Comment (j-vp) [inline src/Vault.sol:693]: `_liquidator` or `liquidator_?`

File: src/Vault.sol [hunk 5]
```diff
@@ -658,92 +648,113 @@ contract Vault {
...
+    }
+
+    /** 
+    @notice Closes the margin account on the vault of the trusted application..
+    @dev The open position is fetched at a contract of the application -> only allow trusted audited protocols!!!
+    @dev Currently only one trusted protocol can be set.
+    */
+    function closeTrustedMarginAccount() public onlyOwner {
+        require(isTrustedProtocolSet, "V_CMA: NOT SET");
+
```
Comment (j-vp) [inline src/Vault.sol:711]: remove whitespace

File: src/Vault.sol [hunk 6]
```diff
@@ -658,92 +648,113 @@ contract Vault {
...
+        require(ITrustedProtocol(trustedProtocol).getOpenPosition(address(this)) == 0, "NON-ZERO OPEN POSITION");
```
Comment (j-vp) [inline src/Vault.sol:712]: Add contract & function abbreviation

File: src/Vault.sol [hunk 7]
```diff
@@ -658,92 +648,113 @@ contract Vault {
...
+
+        isTrustedProtocolSet = false;
```
Comment (j-vp) [inline src/Vault.sol:714]: We should pack this in a struct

File: src/Vault.sol [hunk 8]
```diff
@@ -835,8 +852,8 @@ contract Vault {
     @return success Boolean indicating if there is sufficient free margin to increase the margin position
     @dev All values expressed in the base currency of the vault with same number of decimals as the base currency.
     */
-    function increaseMarginPosition(uint256 baseCurrency, uint256 amount) public onlyAuthorized returns (bool success) {
-        if (baseCurrency != vault._baseCurrency) _setBaseCurrency(baseCurrency);
+    function increaseMarginPosition(address baseCurrency, uint256 amount) public onlyAuthorized returns (bool success) {
+        if (baseCurrency != vault.baseCurrency) _setBaseCurrency(baseCurrency);
         success = getFreeMargin() >= amount;
```
Comment (j-vp) [inline src/Vault.sol:857]: If there is only one protocol that can call this function, why not have "protocol.openposition" included as input in this function? As it's currently implemented this requires another roundtrip

File: src/Vault.sol [hunk 9]
```diff
@@ -846,8 +863,8 @@ contract Vault {
     @dev All values expressed in the base currency of the vault with same number of decimals as the base currency.
     @return success Boolean indicating if there the margin position is successfully decreased.
      */
-    function decreaseMarginPosition(uint256 baseCurrency, uint256) public view onlyAuthorized returns (bool success) {
-        success = baseCurrency == vault._baseCurrency;
+    function decreaseMarginPosition(address baseCurrency, uint256) public view onlyAuthorized returns (bool success) {
```
Comment (j-vp) [inline src/Vault.sol:866]: Should this view function have an auth modifier? Might be a hassle for bots/...

TBD if we should allow the `_setBaseCurrency` in the `increaseMaringPosition` as well, or just return false when the basecurrency isn't the same. Would be better if we can have both functions as public view functions without auth modifiers.

File: src/Vault.sol [hunk 10]
```diff
@@ -881,20 +898,22 @@ contract Vault {
             //higher than 1.15 * 10**57 * 10**18 decimals
             leftHand = totalValue * 100;
             //gas: cannot overflow: uint8 * uint128 << uint256
-            rightHand = uint256(vault._liqThres) * uint256(openDebt);
+            rightHand = uint256(vault.liqThres) * uint256(openDebt);
         }
 
         require(leftHand < rightHand, "This vault is healthy");
```
Comment (j-vp) [inline src/Vault.sol:904]: Function & contract abbrev.

File: src/Vault.sol [hunk 11]
```diff
@@ -862,16 +879,16 @@ contract Vault {
          Sets debtInfo todo: needed?
          Transfers ownership of the proxy vault to the liquidator!
     @param liquidationKeeper Addross of the keeper who initiated the liquidation process.
-    @param liquidator Contract Address of the liquidation logic.
+    @param _liquidator Contract Address of the liquidation logic.
     @return success Boolean returning if the liquidation process is successfully started.
   */
-    function liquidateVault(address liquidationKeeper, address liquidator)
+    function liquidateVault(address liquidationKeeper, address _liquidator)
```
Comment (j-vp) [inline src/Vault.sol:885]: Why do we need a `_liquidator` input if this is already stored on the vault?

File: src/Vault.sol [hunk 12]
```diff
@@ -32,48 +33,32 @@ import "./interfaces/ILiquidityPool.sol";
...
+    address public liquidator;
 
     struct VaultInfo {
-        uint16 _collThres; //2 decimals precision (factor 100)
-        uint8 _liqThres; //2 decimals precision (factor 100)
-        uint8 _baseCurrency;
+        uint16 collThres; //2 decimals precision (factor 100)
+        uint8 liqThres; //2 decimals precision (factor 100)
+        address baseCurrency;
+        //address baseCurrencyAddress;
```
Comment (j-vp) [inline src/Vault.sol:61]: Can we add vaultVersion, life, ... into this struct?

File: src/Vault.sol [hunk 13]
```diff
@@ -881,20 +898,22 @@ contract Vault {
...
                 life,
                 liquidationKeeper,
                 owner,
                 openDebt,
-                vault._liqThres,
-                vault._baseCurrency
+                vault.liqThres,
+                baseCurrencyIdentifier
             ),
             "Failed to start auction!"
```
Comment (j-vp) [inline src/Vault.sol:918]: abbrev.

File: src/Vault.sol (hunk 4 above)
Comment (Thomas-Smets) [reply src/Vault.sol:693]: image
https://docs.soliditylang.org/en/v0.8.16/style-guide.html

File: src/Vault.sol (hunk 8 above)
Comment (Thomas-Smets) [reply src/Vault.sol:857]: Because in the future there should be more functions that can call this function

File: src/Vault.sol (hunk 7 above)
Comment (Thomas-Smets) [reply src/Vault.sol:713]: I propose we do the packing of variables and other gas optimisations after complete refactor (there are some pending PRs after this one)

File: src/Vault.sol (hunk 9 above)
Comment (Thomas-Smets) [reply src/Vault.sol:865]: Because in the future there should be more functions that can call this function

**Review by j-vp — APPROVED** [review]
lgtm

**Pull request merged**